reco_utils/common/tf_utils.py

```diff
@@ -0,0 +1,408 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+import abc
+import itertools
+import tensorflow as tf
+import numpy as np
+import pandas as pd
+
+from reco_utils.common.constants import (
+    DEFAULT_USER_COL,
+    DEFAULT_ITEM_COL
+)
+
+
+MODEL_DIR = 'model_checkpoints'
+
+
+def pandas_input_fn(
+    df,
+    y_col=None,
+    batch_size=128,
+    num_epochs=1,
+    shuffle=False,
+    num_threads=1
+):
+    """Pandas input function for TensorFlow high-level API Estimator.
+
+    tf.estimator.inputs.pandas_input_fn cannot handle array/list column properly.
+    If the df does not include any array/list data column, one can simply use TensorFlow's pandas_input_fn.
+
+    For more information, see (https://www.tensorflow.org/api_docs/python/tf/estimator/inputs/numpy_input_fn)
+
+    Args:
+        df (pd.DataFrame): Data containing features
+        y_col (str): Label column name if df has it.
+        batch_size (int): Batch size for the input function
+        num_epochs (int): Number of epochs to iterate over data. If None will run forever.
+        shuffle (bool): If True, shuffles the data queue.
+        num_threads (int): Number of threads used for reading and enqueueing.
+
+    Returns:
+        (tf.estimator.inputs.numpy_input_fn) function that has signature of ()->(dict of features, targets)
+    """
+
+    X_df = df.copy()
+    if y_col is not None:
+        y = X_df.pop(y_col).values
+    else:
+        y = None
+
+    X = {}
+    for col in X_df.columns:
+        values = X_df[col].values
+        if isinstance(values[0], (list, np.ndarray)):
+            values = np.array([l for l in values], dtype=np.float32)
+        X[col] = values
+
+    input_fn = tf.estimator.inputs.numpy_input_fn(
+        x=X,
+        y=y,
+        batch_size=batch_size,
+        num_epochs=num_epochs,
+        shuffle=shuffle,
+        num_threads=num_threads
+    )
+
+    return input_fn
+
+
+def _build_optimizer(name, lr=0.001, ftrl_l1_reg=0.0):
```

Comment:
these function _build_optimizer, _build_wide_columns, build_model, etc are not tf utils, they are the wide and deep algo. I think we should move this to a new folder in recommenders.

Reply:
_build_wide_columns is wide-deep model specific, but not other functions. 
build_optimizer and build_model can be used any TF Estimators like regular linear regressor or DNNs (w/ TF high-level API). Do we still want to move them?

Reply:
I would say that if it is something specific to W&D, it goes with it. If it is something that can be used by other TF algos, then it should go in a common dir

Reply:
Maybe this can sever you as an example (even though it is work in progress): https://github.com/Microsoft/Recommenders/blob/benchmark_feb/reco_utils/recommender/fastai/fastai_utils.py

Here I'm just refactoring some functions from fastai that we use in the quick start notebook and in the benchmark. I'm doing a minimal thing, not super complicated wrappers over the fastai model

Reply:
Agreed. moved column and model functions to recommender/wide_deep/wide_deep_utils.py and keep rest to the common tf_utils.py